man/all_ref.8
.\" section 8 man page for all_ref.sh
.\"
.\" This man page was first written by Cody Boone Ferguson for the IOCCC
.\" in 2022.
.\"
.\" Humour impairment is not virtue nor is it a vice, it's just plain
.\" wrong: almost as wrong as JSON spec mis-features and C++ obfuscation! :-)
.\"
.\" "Share and Enjoy!"
.\"     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
.\"
.TH all_ref.sh 8 "01 December 2022" "all_ref.sh" "IOCCC tools"
.SH NAME
all_ref.sh \- form \fI*.json.c\fP and \fI*.json.h\fP semantic table files
.SH SYNOPSIS
\fBall_ref.sh\fP [\-h] [\-v level] [\-V] [\-j jsemcgen.sh] info.head.c info.tail.c info.head.h info.tail.h auth.head.c auth.tail.c auth.head.h auth.tail.h info_dir auth_dir ref_dir
.SH DESCRIPTION
This tool forms semantic table files under a given directory which is, in the case of the IOCCC, important for the \fBchkentry(1)\fP tool.
As \fBchkentry(1)\fP is a vitally important part of the IOCCC judging process this tool, along with \fBjsemcgen(8)\fP and \fBjsemtblgen(8)\fP are vital in running the IOCCC but at a lower level.
.PP
The \fIinfo.head.c\fP file is a \fI.info.json\fP style header for .c semantic table files.
.br
The \fIinfo.tail.c\fP file is a \fI.info.json\fP style trailer for .c semantic table files.
.br
The \fIinfo.head.h\fP file is a \fI.info.json\fP style header for .h semantic table files.
.br
The \fIinfo.tail.h\fP file is a \fI.info.json\fP style trailer for .h semantic table files.
.PP
The \fIauth.head.c\fP file is a \fI.auth.json\fP style header for .c semantic table files.
.br
The \fIauth.tail.c\fP file is a \fI.auth.json\fP style trailer for .c semantic table files.
.br
The \fIauth.head.h\fP file is a \fI.auth.json\fP style header for .h semantic table files.
.br
The \fIauth.tail.h\fP file is a \fI.auth.json\fP style trailer for .h semantic table files.
.PP
The \fIinfo_dir\fP directory is expected to have \fI*.json\fP files as \fI.info.json\fP files.
.br
The \fIauth_dir\fP directory is expected to have \fI*.json\fP files as \fI.auth.json\fP files.
.br
The \fIref_dir\fP directory is a subdirectory under which the semantic table files are written.
.SH OPTIONS
.TP
\fB\-h\fP
Show help and exit.
.TP
\fB\-v \fIlevel\fP\fP
Set verbosity level to level.
.TP
\fB\-V\fP
Show version and exit.
.TP
\fB-j \fIjsemcgen.sh\fP\fP
Set path to \fBjsemcgen(8)\fP.
.SH EXIT STATUS
.TP
0
all is well
.TQ
2
\fB\-h\fP or \fB\-V\fP used and help string printed or version string printed
.TQ
3
Command line error
.TQ
4
\fBjsemcgen.sh\fP not found or not executable
.TQ
5
Missing or not readable header or trailer file
.TQ
6
Missing, not readable, or not writable \fIinfo_dir\fP, \fIauth_dir\fP and/or \fIref_dir\fP
.TQ
>= 10
Internal error
.SH NOTES
.PP
The JSON parser \fBjparse\fP was co\-developed by Cody Boone Ferguson and Landon Curt Noll (one of the IOCCC Judges) in support for IOCCCMOCK, IOCCC28 and beyond.
.PP
This tool is needed in order to generate the JSON semantics table for \fBchkentry(1)\fP.
.SH BUGS
If you have a problem with the tool (not JSON itself! :\-) ) you can report it at the GitHub issues page.
It can be found at
.br
\fI\<https://github.com/ioccc\-src/mkiocccentry/issues\>\fP.
.SH SEE ALSO
\fBjsemtblgen(8)\fP, \fBchkentry(1)\fP, \fBjsemcgen(8)\fP


CHANGES.md
# Major changes to the IOCCC entry toolkit


## Release 0.8.1 2022-11-19

This is "roll-up" release was added so that it could be referenced
in the https://www.ioccc.org/index.html news section.

Updated legacy_clobber rule to remove older directories no longer used.

Moved test directory to test_ioccc to avoid Makefile confusion
with the test rule.

Restructure dbg facility and Makefile in accordance with
the new method of copying code from external GitHub repositories.

Improve formatting with the hostchk_warning rule.

Added initial dyn_array.md documentation.

Typo fixes to the verge.8 man page.

Minor fixes to jparse.md.

Updated txzchk comments (source, header, man page).

Improved and updated README.md documentation.

Improved bug_report.sh: in particular in reporting versions.

Bug fixes and improvements to several shell scripts.
Made changes to satisfy shellcheck as well.

Added rules in the Makefile to make is easier to import code from
external GitHub repositories.

```
This repo is designed to be a standalone repo.  Even though we use other
repositories, we prefer to NOT clone them.  We want this repo to depend
on a specific version of such code so that a change in the code of those
external repositories will NOT impact this repo.

For that reason, and others, we maintain a private copy of an external
repository as clone.repo.  The clone.repo directory is excluded from
out repo via the .gitignore file.  We copy clone.repo/ into repo/ and
check in those file directly into this repo.
```

The dbg code is now a copy of the code from the dbg repo.

Update JSON parser version.  Changed JSON_PARSER_VERSION from "0.11
2022-11-04" to "0.12 2022-11-09".  There was an API change (an
addition of a filename argument in some calls) that suggested version change.


## Release 0.8 2022-11-09

Improve the situation with some systems requiring some feature test macros.
Removed all references to `${TIMEGM_PROBLEM}` in Makefile.
Removed all references to `${D_LEGACY}` in Makefile.
Cleaned out the `$(WARN_FLAGS}` list (from legacy hosts).
These changes allow for cleanly compiling under CentOS without as many
complexities (it was also important for modern systems like Fedora).

Made a number of important fixes to `chkentry`.

Made a number of important fixes to JSON parser.
Pre-scan documents and strings for problematic bytes prior to trying to parse it
as json.

The JSON scanner and parser are now re-entrant!

Improved JSON parser error messages by adding filename and location (lines and
columns).

Added filename argument to top level parser functions:

```c
extern struct json *parse_json(char const *ptr, size_t len, char const *filename, bool *is_valid);
extern struct json *parse_json_stream(FILE *stream, char const *filename, bool *is_valid);
```

This was one part of making the scanner and parser re-entrant. See the git log
and diffs of `jparse.l` and `jparse.y` for more details.

The documentation of the JSON parser is not complete and it will not be
completed until after IOCCCMOCK.

Fixed several memory leaks.

Fixed a number of typos in comments.

Many man pages updated and/or fixed.
Removed `limit_ioccc(8)` man page.

Made a number of shell script fixes with respect to the `mktemp(1)` command.

Experimenting with a system independent `make depend` tool: work in progress.


## Release 0.7 2022-11-05

Released IOCCC entry toolkit v0.7 2022-11-05

Changed BUG_REPORT_VERSION from "0.5 2022-11-03" to "0.6 2022-11-06".
Changed FMT_DEPEND_VERSION from "1.0 2022-09-28" to "1.1 2022-11-05".
Changed CHKENTRY_TEST_VERSION from "0.1 2022-10-11" to "0.2 2022-11-04".
Changed IOCCC_TEST_VERSION from "0462202-04-23" (sic) to "0.5 2022-11=04".
Changed TEST_IOCCCSIZE_VERSION from "1.1 2022-10-16" to ="1.2 2022-11-04".
Changed JPARSE_TEST_VERSION from "0.3 2022-10-22" to "0.4 2022-11-04".
Changed JSTR_TEST_VERSION from "0.4 2022-09-28" to "0.4 2022-11-05".
Changed TXZCHK_TEST_VERSION from "0.4 2022-10-20" to "0.5 2022-11-04".

The `chkentry(1)` tool is now code complete!  Added test suite for
`chkentry` called test/chkentry_test.sh.  The `chkentry(1)` tool
passes all tests.

Removed prefix in bison and flex code. The programmer's apology and
warning are still in `sorry.tm.ca.h`.

A test suite for `txzchk` was added.

Removed the `MAX_DIR_KSIZE` limit. There still is a maximum tarball
size but now there also is a limit of the number of files in the
tarball. `txzchk` and `mkiocccentry` were updated for this.

We're one step closer to being able to invite the public to
review these tools.

Created the test sub-directory.  Moved many test tools under the
test sub-directory.  Moved log files under test.

Created the man sub-directory.  Moved all man pages under the
man sub-directory.  The html files are also build under man.

Added the soup sub-directory.  Moved tools related to building semantic
test tables for chkentry under the soup sub-directory.  Moved ref
under soup.

Added fmt_depend.sh under the soup sub-directory to format
the make depend lines in a system independent way.

Added -Z topdir flags to soup/fmt_depend.sh, test/ioccc_test.sh,
test/jstr_test.sh, test/mkiocccentry_test.sh, and test/txzchk_test.sh.
By default these tools cd to the top level directory where the
mkiocccentry.c source file resides.  By use of -Z topdir, one can
force the top level directory location.

Experimenting with compiling code in sub-directories such as soup
and test.


## Release 0.6 2022-09-02

Released IOCCC entry toolkit v0.6 2022-09-02

Updated CHANGES.md for v0.6 2022-09-02.

Changed `MIN_TIMESTAMP` from 1655958810 to 1662145368.

Updated `mkiocccentry` from "0.40 2022-03-15" to "0.41 2022-09-02".
Updated `.info.json` version from "1.10 2022-06-22" to "1.11 2022-09-02".
Updated `.author.json` version from "1.13 2022-06-22" to "1.14 2022-09-02".

Improved code to use new facilities for output to a buffer from dbg release of
v2.5 2022-07-23.


soup/all_ref.sh
#!/usr/bin/env bash
#
# all_ref - form the ref directory
#
# For the ref sub-directory using via ./jsemcgen.sh, from test .info.json
# and .auth.json files, form *.json.c and *.json.h semantic table files
# under a given directory.
#
# The concept of this file was developed by:
#
#	chongo (Landon Curt Noll, http://www.isthe.com/chongo/index.html) /\oo/\
#
# The JSON parser was co-developed by:
#
#	@xexyl
#	https://xexyl.net		Cody Boone Ferguson
#	https://ioccc.xexyl.net
# and:
#	chongo (Landon Curt Noll, http://www.isthe.com/chongo/index.html) /\oo/\
#
# Share and enjoy! :-)
#     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
#

# setup
#
export V_FLAG="0"
export JSEMCGEN_SH="./jsemcgen.sh"
export ALL_REF_VERSION="1.0 2022-10-30"

export USAGE="usage: $0 [-h] [-v level] [-V] [-j jsemcgen.sh] info.head.c info.tail.c info.head.h
	info.tail.h auth.head.c auth.tail.c auth.head.h auth.tail.h info_dir auth_dir ref_dir

	-h		print help message and exit
	-v level	set verbosity level (def level: 0)
	-V		print version string and exit

	-j jsemcgen.sh	path to jsemcgen.sh (def: ./jsemcgen.sh)

	info.head.c	.info.json style header for .c semantic table files
	info.tail.c	.info.json style trailer for .c semantic table files
	info.head.h	.info.json style header for .h semantic table files
	info.tail.h	.info.json style trailer for .h semantic table files

	auth.head.c	.auth.json style header for .c semantic table files
	auth.tail.c	.auth.json style trailer for .c semantic table files
	auth.head.h	.auth.json style header for .h semantic table files
	auth.tail.h	.auth.json style trailer for .h semantic table files

	info_dir	process *.json files as .info.json files
	auth_dir	process *.json files as .auth.json files

	ref_dir		sub-directory under which semantic table files are written

Exit codes:
     0	 all is well
     2	 -h and help string printed or -V and version string printed
     3	 command line error
     4	 jsemcgen.sh not found or not executable
     5	 missing or not readable header or trailer file
     6	 missing, not readable, or not writable info_dir, auth_dir and/or ref_dir
 >= 10	 internal error"

# parse args
#
while getopts :hv:Vj: flag; do
    case "$flag" in
    h) echo "$USAGE" 1>&2
       exit 2
       ;;
    v) V_FLAG="$OPTARG";
       JSEMTBLGEN_ARGS="$JSEMTBLGEN_ARGS -v '$V_FLAG'";
       ;;
    V) echo "$ALL_REF_VERSION"
       exit 2
       ;;
    j) JSEMCGEN_SH="$OPTARG";
       ;;
    \?) echo "$0: ERROR: invalid option: -$OPTARG" 1>&2
       exit 3
       ;;
    :) echo "$0: ERROR: option -$OPTARG requires an argument" 1>&2
       exit 3
       ;;
   *)
       ;;
    esac
done

# check args
#
shift $(( OPTIND - 1 ));
if [[ $# -ne 11 ]]; then
    echo "$0: ERROR: expected 11 arguments, found $#" 1>&2
    echo "$USAGE" 1>&2
    exit 3
fi
INFO_HEAD_C="$1"
INFO_TAIL_C="$2"
INFO_HEAD_H="$3"
INFO_TAIL_H="$4"
AUTH_HEAD_C="$5"
AUTH_TAIL_C="$6"
AUTH_HEAD_H="$7"
AUTH_TAIL_H="$8"
INFO_DIR="$9"
AUTH_DIR="${10}"
REF_DIR="${11}"
export INFO_HEAD_C INFO_TAIL_C INFO_HEAD_H INFO_TAIL_H
export AUTH_HEAD_C AUTH_TAIL_C AUTH_HEAD_H AUTH_TAIL_H
export AUTH_DIR AUTH_DIR REF_DIR
if [[ $V_FLAG -ge 3 ]]; then
    echo "$0: debug[3]: INFO_HEAD_C=$INFO_HEAD_C" 1>&2
    echo "$0: debug[3]: INFO_TAIL_C=$INFO_TAIL_C" 1>&2
    echo "$0: debug[3]: INFO_HEAD_H=$INFO_HEAD_H" 1>&2
    echo "$0: debug[3]: INFO_TAIL_H=$INFO_TAIL_H" 1>&2
    echo "$0: debug[3]: AUTH_HEAD_C=$AUTH_HEAD_C" 1>&2
    echo "$0: debug[3]: AUTH_TAIL_C=$AUTH_TAIL_C" 1>&2
    echo "$0: debug[3]: AUTH_HEAD_H=$AUTH_HEAD_H" 1>&2
    echo "$0: debug[3]: AUTH_TAIL_H=$AUTH_TAIL_H" 1>&2
    echo "$0: debug[3]: INFO_DIR=$INFO_DIR" 1>&2
    echo "$0: debug[3]: AUTH_DIR=$AUTH_DIR" 1>&2
    echo "$0: debug[3]: REF_DIR==$REF_DIR" 1>&2
fi

# firewall
#
if [[ ! -x $JSEMCGEN_SH ]]; then
    echo "$0: ERROR: jsemcgen.sh is not an executable: $JSEMCGEN_SH" 1>&2
    exit 4
fi
#
if [[ ! -e $INFO_HEAD_C ]]; then
    echo "$0: ERROR: missing file info.head.c: $INFO_HEAD_C" 1>&2
    exit 5
fi
if [[ ! -f $INFO_HEAD_C ]]; then
    echo "$0: ERROR: info.head.c not a file: $INFO_HEAD_C" 1>&2
    exit 5
fi
if [[ ! -r $INFO_HEAD_C ]]; then
    echo "$0: ERROR: info.head.c not a readable file: $INFO_HEAD_C" 1>&2
    exit 5
fi
#
if [[ ! -e $INFO_HEAD_H ]]; then
    echo "$0: ERROR: missing file info.head.h: $INFO_HEAD_H" 1>&2
    exit 5
fi
if [[ ! -f $INFO_HEAD_H ]]; then
    echo "$0: ERROR: info.head.h not a file: $INFO_HEAD_H" 1>&2
    exit 5
fi
if [[ ! -r $INFO_HEAD_H ]]; then
    echo "$0: ERROR: info.head.h not a readable file: $INFO_HEAD_H" 1>&2
    exit 5
fi
#
if [[ ! -e $INFO_TAIL_C ]]; then
    echo "$0: ERROR: missing file info.tail.c: $INFO_TAIL_C" 1>&2
    exit 5
fi
if [[ ! -f $INFO_TAIL_C ]]; then
    echo "$0: ERROR: info.tail.c not a file: $INFO_TAIL_C" 1>&2
    exit 5
fi
if [[ ! -r $INFO_TAIL_C ]]; then
    echo "$0: ERROR: info.tail.c not a readable file: $INFO_TAIL_C" 1>&2
    exit 5
fi
#
if [[ ! -e $INFO_TAIL_H ]]; then
    echo "$0: ERROR: missing file info.tail.h: $INFO_TAIL_H" 1>&2
    exit 5
fi
if [[ ! -f $INFO_TAIL_H ]]; then
    echo "$0: ERROR: info.tail.h not a file: $INFO_TAIL_H" 1>&2
    exit 5
fi
if [[ ! -r $INFO_TAIL_H ]]; then
    echo "$0: ERROR: info.tail.h not a readable file: $INFO_TAIL_H" 1>&2
    exit 5
fi
#
if [[ ! -e $AUTH_HEAD_C ]]; then
    echo "$0: ERROR: missing file auth.head.c: $AUTH_HEAD_C" 1>&2
    exit 5
fi
if [[ ! -f $AUTH_HEAD_C ]]; then
    echo "$0: ERROR: auth.head.c not a file: $AUTH_HEAD_C" 1>&2
    exit 5
fi
if [[ ! -r $AUTH_HEAD_C ]]; then
    echo "$0: ERROR: auth.head.c not a readable file: $AUTH_HEAD_C" 1>&2
    exit 5
fi
#
if [[ ! -e $AUTH_HEAD_H ]]; then
    echo "$0: ERROR: missing file auth.head.h: $AUTH_HEAD_H" 1>&2
    exit 5
fi
if [[ ! -f $AUTH_HEAD_H ]]; then
    echo "$0: ERROR: auth.head.h not a file: $AUTH_HEAD_H" 1>&2
    exit 5
fi
if [[ ! -r $AUTH_HEAD_H ]]; then
    echo "$0: ERROR: auth.head.h not a readable file: $AUTH_HEAD_H" 1>&2
    exit 5
fi
#
if [[ ! -e $AUTH_TAIL_C ]]; then
    echo "$0: ERROR: missing file auth.tail.c: $AUTH_TAIL_C" 1>&2
    exit 5
fi
if [[ ! -f $AUTH_TAIL_C ]]; then
    echo "$0: ERROR: auth.tail.c not a file: $AUTH_TAIL_C" 1>&2
    exit 5
fi
if [[ ! -r $AUTH_TAIL_C ]]; then
    echo "$0: ERROR: auth.tail.c not a readable file: $AUTH_TAIL_C" 1>&2
    exit 5
fi
#
if [[ ! -e $AUTH_TAIL_H ]]; then
    echo "$0: ERROR: missing file auth.tail.h: $AUTH_TAIL_H" 1>&2
    exit 5
fi
if [[ ! -f $AUTH_TAIL_H ]]; then
    echo "$0: ERROR: auth.tail.h not a file: $AUTH_TAIL_H" 1>&2
    exit 5
fi
if [[ ! -r $AUTH_TAIL_H ]]; then
    echo "$0: ERROR: auth.tail.h not a readable file: $AUTH_TAIL_H" 1>&2
    exit 5
fi
#
if [[ ! -e $INFO_DIR ]]; then
    echo "$0: ERROR: missing directory info_dir: $INFO_DIR" 1>&2
    exit 6
fi
if [[ ! -d $INFO_DIR ]]; then
    echo "$0: ERROR: info_dir not a directory: $INFO_DIR" 1>&2
    exit 6
fi
if [[ ! -w $INFO_DIR ]]; then
    echo "$0: ERROR: info_dir not a writable directory: $INFO_DIR" 1>&2
    exit 6
fi
if [[ ! -r $INFO_DIR ]]; then
    echo "$0: ERROR: info_dir not a readable directory: $INFO_DIR" 1>&2
    exit 6
fi
#
if [[ ! -e $AUTH_DIR ]]; then
    echo "$0: ERROR: missing directory auth_dir: $AUTH_DIR" 1>&2
    exit 6
fi
if [[ ! -d $AUTH_DIR ]]; then
    echo "$0: ERROR: auth_dir not a directory: $AUTH_DIR" 1>&2
    exit 6
fi
if [[ ! -w $AUTH_DIR ]]; then
    echo "$0: ERROR: auth_dir not a writable directory: $AUTH_DIR" 1>&2
    exit 6
fi
if [[ ! -r $AUTH_DIR ]]; then
    echo "$0: ERROR: auth_dir not a readable directory: $INFO_DIR" 1>&2
    exit 6
fi
#
if [[ ! -e $REF_DIR ]]; then
    echo "$0: ERROR: missing directory ref_dir: $REF_DIR" 1>&2
    exit 6
fi
if [[ ! -d $REF_DIR ]]; then
    echo "$0: ERROR: ref_dir not a directory: $REF_DIR" 1>&2
    exit 6
fi
if [[ ! -w $REF_DIR ]]; then
    echo "$0: ERROR: ref_dir not a writable directory: $REF_DIR" 1>&2
    exit 6
fi
if [[ ! -r $REF_DIR ]]; then
    echo "$0: ERROR: ref_dir not a readable directory: $INFO_DIR" 1>&2
    exit 6
fi

TMP_FILE=$(mktemp .all_ref.XXXXXXXXXX.out)
status="$?"
if [[ $status -ne 0 ]]; then
    echo "$0: ERROR: mktemp .all_ref.XXXXXXXXXX exit code: $status" 1>&2
    exit 10
fi
trap "rm -f \$TMP_FILE; exit" 0 1 2 3 15
export TMP_FILE

# form .info.json style semantic table .c files
#
if [[ $V_FLAG -ge 1 ]]; then
    echo "$0: debug[1]: form .info.json style semantic table .c files from $INFO_DIR" 1>&2
fi
find "$INFO_DIR" -type f -name '*.json' -print 2>/dev/null | while read -r path; do

    # determine filenames
    #
    file=$(basename "$path")
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: processing .info.json style file: $file" 1>&2
    fi
    out="$REF_DIR/$file.c"
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: will try to form: $out" 1>&2
    fi

    # form output file
    #
    if [[ $V_FLAG -ge 5 ]]; then
	echo "$0: debug[5]: about to execute: $JSEMCGEN_SH -N sem_info -P chk -- $path . . . $TMP_FILE" 1>&2
    fi
    "$JSEMCGEN_SH" -N sem_info -P chk -- "$path" . . . "$TMP_FILE"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: $JSEMCGEN_SH -N sem_info -P chk -- $path . . . $TMP_FILE  exit code: $status" 1>&2
	exit 11
    fi
    if [[ $V_FLAG -ge 5 ]]; then
	echo "$0: debug[5]: about to execute: cat $INFO_HEAD_C $TMP_FILE $INFO_TAIL_C > $out" 1>&2
    fi
    cat "$INFO_HEAD_C" "$TMP_FILE" "$INFO_TAIL_C" > "$out"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: cat $INFO_HEAD_C $TMP_FILE $INFO_TAIL_C > $out exit code: $status" 1>&2
	exit 12
    fi
done

# form .info.json style semantic table .h files
#
if [[ $V_FLAG -ge 1 ]]; then
    echo "$0: debug[1]: form .info.json style semantic table .h files from $INFO_DIR" 1>&2
fi
find "$INFO_DIR" -type f -name '*.json' -print 2>/dev/null | while read -r path; do

    # determine filenames
    #
    file=$(basename "$path")
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: processing .info.json style file: $file" 1>&2
    fi
    out="$REF_DIR/$file.h"
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: will try to form: $out" 1>&2
    fi

    # form output file
    #
    if [[ $V_FLAG -ge 5 ]]; then
	echo "$0: debug[5]: about to execute: $JSEMCGEN_SH -N sem_info -P chk -I -- $path . . . $TMP_FILE" 1>&2
    fi
    "$JSEMCGEN_SH" -N sem_info -P chk -I -- "$path" . . . "$TMP_FILE"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: $JSEMCGEN_SH -N sem_info -P chk -I -- $path . . . $TMP_FILE  exit code: $status" 1>&2
	exit 13
    fi
    if [[ $V_FLAG -ge 5 ]]; then
	echo "$0: debug[5]: about to execute: cat $INFO_HEAD_H $TMP_FILE $INFO_TAIL_H > $out" 1>&2
    fi
    cat "$INFO_HEAD_H" "$TMP_FILE" "$INFO_TAIL_H" > "$out"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: cat $INFO_HEAD_H $TMP_FILE $INFO_TAIL_H > $out exit code: $status" 1>&2
	exit 14
    fi
done

# form .auth.json style semantic table .c files
#
if [[ $V_FLAG -ge 1 ]]; then
    echo "$0: debug[1]: form .auth.json style semantic table .c files from $AUTH_DIR" 1>&2
fi
find "$AUTH_DIR" -type f -name '*.json' -print 2>/dev/null | while read -r path; do

    # determine filenames
    #
    file=$(basename "$path")
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: processing .auth.json style file: $file" 1>&2
    fi
    out="$REF_DIR/$file.c"
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: will try to form: $out" 1>&2
    fi

    # form output file
    #
    if [[ $V_FLAG -ge 5 ]]; then
	echo "$0: debug[5]: about to execute: $JSEMCGEN_SH -N sem_auth -P chk -- $path . . . $TMP_FILE" 1>&2
    fi
    "$JSEMCGEN_SH" -N sem_auth -P chk -- "$path" . . . "$TMP_FILE"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: $JSEMCGEN_SH -N sem_auth -P chk -- $path . . . $TMP_FILE  exit code: $status" 1>&2
	exit 15
    fi
    if [[ $V_FLAG -ge 5 ]]; then
	echo "$0: debug[5]: about to execute: cat $AUTH_HEAD_C $TMP_FILE $AUTH_TAIL_C > $out" 1>&2
    fi
    cat "$AUTH_HEAD_C" "$TMP_FILE" "$AUTH_TAIL_C" > "$out"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: cat $AUTH_HEAD_C $TMP_FILE $AUTH_TAIL_C > $out exit code: $status" 1>&2
	exit 16
    fi
done

# form .auth.json style semantic table .h files
#
if [[ $V_FLAG -ge 1 ]]; then
    echo "$0: debug[1]: form .auth.json style semantic table .h files from $AUTH_DIR" 1>&2
fi
find "$AUTH_DIR" -type f -name '*.json' -print 2>/dev/null | while read -r path; do

    # determine filenames
    #
    file=$(basename "$path")
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: processing .auth.json style file: $file" 1>&2
    fi
    out="$REF_DIR/$file.h"
    if [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: will try to form: $out" 1>&2
    fi

    # form output file
    #
    if [[ $V_FLAG -ge 5 ]]; then
	echo "$0: debug[5]: about to execute: $JSEMCGEN_SH -N sem_auth -P chk -I -- $path . . . $TMP_FILE" 1>&2
    fi
    "$JSEMCGEN_SH" -N sem_auth -P chk -I -- "$path" . . . "$TMP_FILE"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: $JSEMCGEN_SH -N sem_auth -P chk -I -- $path . . . $TMP_FILE  exit code: $status" 1>&2
	exit 17
    fi
    if [[ $V_FLAG -ge 5 ]]; then
	echo "$0: debug[5]: about to execute: cat $AUTH_HEAD_H $TMP_FILE $AUTH_TAIL_H > $out" 1>&2
    fi
    cat "$AUTH_HEAD_H" "$TMP_FILE" "$AUTH_TAIL_H" > "$out"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: cat $AUTH_HEAD_H $TMP_FILE $AUTH_TAIL_H > $out exit code: $status" 1>&2
	exit 18
    fi
done

# All Done!!! -- Jessica Noll, Age 2
#
rm -f "$TMP_FILE"
exit 0
